### PR TITLE
Relax ROCm restrictions with LLVM 21

### PR DESCRIPTION
**Pull request by jabraham17**
Relaxes the restriction to use ROCm with a bundled LLVM.

The new state as of this PR

* With ROCm 6.3, you can use either the bundled LLVM or LLVM 21+
* With ROCm <6.3, you can still only use the bundled LLVM
    * This is a LLVM limitation, they require ROCm 6.3+

Note: I only ran the `COMM=none` test suite, which did not pass 100%. Most tests passed, a few edge cases (like interop printf) did not work quite right. I consider that future work. IMO, its more important for things to mostly work and we can improve specific cases as users run into issues.

Resolves https://github.com/chapel-lang/chapel/issues/28007

[Reviewed by @e-kayrakli]

### CODE REVIEW

File: compiler/llvm/clangUtil.cpp [hunk 1]
```diff
@@ -3029,6 +3029,17 @@ static void helpComputeClangArgs(std::string& clangCC,
   }
 #endif
 
+#if HAVE_LLVM_VER >= 210
+  if (usingGpuLocaleModel() &&
+      getGpuCodegenType() == GpuCodegenType::GPU_CG_AMD_HIP) {
+    // this is need to make hipcub wrappers work since
+    // TempStorage is marked deprecated, but I can't seem to find an alternative
+    // the rocm docs do not mention the deprecation and I can't find proper docs
+    // for the new API that the warning refers to
```
Comment (e-kayrakli) [inline compiler/llvm/clangUtil.cpp:3038]: :)

File: runtime/include/gpu/amd/chpl-gpu-dev-reduce.h [hunk 2]
```diff
@@ -91,4 +91,8 @@ GPU_DEV_REDUCE(DEF_ONE_DEV_REDUCE_RET_VAL, Max, max);
 
 #undef DEF_ONE_DEV_REDUCE_RET_VAL
 
+#if (ROCM_VERSION_MAJOR == 6 && ROCM_VERSION_MINOR >= 3) || ROCM_VERSION_MAJOR > 6
+#pragma clang diagnostic pop
```
Comment (e-kayrakli) [inline runtime/include/gpu/amd/chpl-gpu-dev-reduce.h]: Just for my edification, what's this about?

**Review by e-kayrakli — APPROVED** [review]

File: runtime/include/gpu/amd/chpl-gpu-dev-reduce.h (hunk 2 above)
Comment (jabraham17) [reply runtime/include/gpu/amd/chpl-gpu-dev-reduce.h]: Its not supposed to be here :).

leftover artifact from [before](https://github.com/chapel-lang/chapel/pull/28421/changes/14e8fee8a2ffe4a1755d5a2a28370a0cd8caf464), which was reverted in https://github.com/chapel-lang/chapel/pull/28421/changes/a342834a9b4112cf150e640bbdda18c28a359031

**Comment by arifthpe** [comment]
should this change be reflected in the Spack package's requirements on next release?

**Comment by jabraham17** [comment]
Yes!
